File: frontend/lib/account/models/account.dart
import 'package:flutter/material.dart';
import 'package:sprout/account/models/institution.dart';

enum DepositoryAccountType {
  savings('Savings'),
  checking('Checking'),
  hysa('HYSA');

  const DepositoryAccountType(this.value);
  final String value;
}

enum InvestmentAccountType {
  four01k('401K'),
  brokerage('Brokerage'),
  ira('IRA'),
  hsa('HSA');

  const InvestmentAccountType(this.value);
  final String value;
}

enum LoanAccountType {
  student('Student'),
  mortgage('Mortgage'),
  personal('Personal'),
  auto('Auto');

  const LoanAccountType(this.value);
  final String value;
}

enum CreditAccountType {
  travel('Travel'),
  cashBack('Cash Back');

  const CreditAccountType(this.value);
  final String value;
}

/// This account defines what an account should look like from the API
class Account {
  final String id;
  final String name;
  final String provider;
  final String currency;
  final String type;

  final double balance;
  final double availableBalance;

  final Institution institution;

  /// An icon to use in the event the institution url's logo can't be found
  final IconData fallbackIcon;

  /// The subtype of this account. For example, a depository could be a checking account, savings account, or HYSA.
  String? subType;

  Account({
    required this.id,
    required this.name,
    required this.balance,
    required this.fallbackIcon,
    required this.provider,
    required this.availableBalance,
    required this.currency,
    required this.type,
    this.subType,
    required this.institution,
  });

  factory Account.fromJson(Map<String, dynamic> json) {
    IconData defaultIcon = Icons.account_balance;
    if (json['type'] == 'depository') {
      defaultIcon = Icons.account_balance;
    } else if (json['type'] == 'loan') {
      defaultIcon = Icons.savings;
    } else if (json['type'] == 'credit') {
      defaultIcon = Icons.credit_card;
    } else if (json['type'] == 'investment') {
      defaultIcon = Icons.trending_up;
    }

    return Account(
      id: json['id'] as String,
      name: json['name'] as String,
      provider: json['provider'] as String,
      currency: json['currency'] as String,
      type: json['type'] as String,
      balance: (json['balance'] as num).toDouble(),
      availableBalance: (json['availableBalance'] as num).toDouble(),
      subType: json['subType'],
      fallbackIcon: defaultIcon,
      institution: Institution.fromJson(json["institution"]),
    );
  }

  /// Converts this object to JSON object
  dynamic toJson() {
    return {
      'id': id,
      'name': name,
      'provider': provider,
      'currency': currency,
      'type': type,
      'balance': balance,
      'availableBalance': availableBalance,
      'subType': subType,
    };
  }

  /// Returns if this account affects the net worth negativity due to being a loan type.
  get isNegativeNetWorth {
    return type == "credit" || type == "loan";
  }
}

File: backend/src/account/model/account.ts
import { AccountHistory } from "@backend/account/model/account.history";
import { CreditAccountType, CryptoAccountType, DepositoryAccountType, InvestmentAccountType, LoanAccountType } from "@backend/account/model/account.type";
import { DatabaseDecorators } from "@backend/database/decorators";
import { DatabaseBase } from "@backend/database/model/database.base";
import { Institution } from "@backend/institution/model/institution";
import { User } from "@backend/user/model/user";
import { ManyToOne } from "typeorm";

@DatabaseDecorators.entity()
export class Account extends DatabaseBase {
  @DatabaseDecorators.column({ nullable: false, unique: true })
  name: string;

  /** Where this account came from */
  @DatabaseDecorators.column({ nullable: false })
  provider: "simple-fin";

  /** The institution associated to this account */
  @ManyToOne(() => Institution, (i) => i.id, { eager: true, cascade: true })
  institution: Institution;

  /** The user this account belongs to */
  @ManyToOne(() => User, (u) => u.id)
  user: User;

  /** The currency this account uses */
  @DatabaseDecorators.column({ nullable: false })
  currency: string;

  /** The current balance of the account */
  @DatabaseDecorators.numericColumn({ nullable: false })
  balance: number;
  /** The available balance to this account */
  @DatabaseDecorators.numericColumn({ nullable: false })
  availableBalance: number;

  /** The type of this account to better separate it from the others. */
  @DatabaseDecorators.column({ nullable: false, type: "varchar" })
  type: "depository" | "credit" | "loan" | "investment" | "crypto";

  /** The subtype of this account. For example, a depository could be a checking account, savings account, or HYSA. */
  @DatabaseDecorators.column({ nullable: true, type: "varchar" })
  subType?: DepositoryAccountType | InvestmentAccountType | LoanAccountType | CreditAccountType | CryptoAccountType;

  /** Any extra data that we want to store as JSON */
  @DatabaseDecorators.jsonColumn({ nullable: true })
  extra?: object;

  /** Given a user, returns all accounts in the database for that user */
  static getForUser(user: User) {
    return Account.find({ where: { user } });
  }

  constructor(
    name: string,
    provider: Account["provider"],
    user: User,
    institution: Institution,
    balance: number,
    availableBalance: number,
    type: Account["type"],
    currency: string,
  ) {
    super();
    this.name = name;
    this.provider = provider;
    this.user = user;
    this.institution = institution;
    this.balance = balance;
    this.availableBalance = availableBalance;
    this.type = type;
    this.currency = currency;
  }

  /** Turns this account to act like account history for today */
  toAccountHistory(date = new Date()) {
    return AccountHistory.fromPlain({
      balance: this.balance,
      account: this,
      availableBalance: this.availableBalance,
      time: date,
    });
  }

  /** Returns if this account affects the net worth negativity due to being a loan type. */
  get isNegativeNetWorth() {
    return this.type === "credit" || this.type === "loan";
  }

  /** Validates that the given sub-type exists in an enum. Throws an error if it doesn't. */
  static validateSubType(subType: string) {
    // Validate that this sub type exists
    const allSubTypes = [
      ...Object.values(DepositoryAccountType),
      ...Object.values(InvestmentAccountType),
      ...Object.values(LoanAccountType),
      ...Object.values(CreditAccountType),
      ...Object.values(CryptoAccountType),
    ];
    if (!allSubTypes.includes(subType as any)) throw new Error(`Invalid subType provided: ${subType}`);
  }
}
